### PR TITLE
Print the value's type help for undocumented bindings

### LINKED ISSUE #13002: Print the type's help for undocumented bindings
In 0.3:

```txt
julia> sparse([1,2,3]);

help?> ans

	[1, 1]  =  1
	[2, 1]  =  2
	[3, 1]  =  3 is of type
DataType   : SparseMatrixCSC{Int64,Int64} (constructor with 1 method)
  supertype: AbstractSparseArray{Int64,Int64,2}
  fields   : (:m,:n,:colptr,:rowval,:nzval)
```

Currently:

```txt
help?> ans
search: ans transpose transpose! ctranspose ctranspose! randsubseq randstring
```
This was a nice way to explore an API.  It'd be nice for an undocumented binding/object to print out its type and the type's help (which is much nicer now with #12996).

### PR DESCRIPTION
**Pull request by jakebolewski**
exclude macros as this is just exposing their internals

closes #13002

cc @mbauman

### CODE REVIEW

**Comment by MichaelHatherly** [comment]
:+1:

Seems like a missed opportunity not doing anything for undocumented macros. Perhaps displaying their signatures would be helpful:

```
help?> @foo

Do documentation found.

@foo(a, b, c...)
```

**Comment by jakebolewski** [comment]
@MichaelHatherly I added a fallback for the macro case as well if you want to test it out.

**Comment by MichaelHatherly** [comment]
Thanks, works great! Tests will fail since https://github.com/JuliaLang/julia/pull/13013/files#diff-3f4eeb007d2f33142bda5be1dc90da75R240 doesn't return ``nothing`` anymore.

**Comment by mbauman** [comment]
This is great.  Thanks Jake.

Could we also add a bit of meta-information about what the help system is doing?  E.g., `catdoc(md"`$b` is of type `$(typeof(v))`:", typesummary(typeof(v)))`.  Maybe we should also use `doc(typeof(v))` instead of `typesummary` so that way (hopefully) more helpful docs show up if they're defined.

File: base/docs/Docs.jl [hunk 1]
```diff
@@ -95,9 +95,49 @@ function doc(obj)
...
+        push!(parts, string("```julia\n", takebuf_string(io), "\n```"))
+    end
+    Markdown.parse(join(parts,'\n'))
+end
+
 function doc(b::Binding)
     d = invoke(doc, Tuple{Any}, b)
-    d == nothing ? doc(getfield(b.mod, b.var)) : d
+    if d == nothing
+        v = getfield(b.mod,b.var)
```
Comment (mbauman) [inline base/docs/Docs.jl]: Is there any chance that this might not be defined?

Comment (jakebolewski) [inline base/docs/Docs.jl]: an error will be thrown in the `Binding` code

```juia
help?> Main.a
ERROR: "a" is not defined in module Main
 in error at ./error.jl:21
 in which_module at ./reflection.jl:315
 in call at ./docs/bindings.jl:8
 in eval_user_input at REPL.jl:64
 [inlined code] from REPL.jl:93
 in anonymous at task.jl:68
```

**Comment by jakebolewski** [comment]
@mbauman good suggestion about meta information, I'll add that.

**Comment by mbauman** [comment]
Functions could probably use a special case, too:

```
julia> undocumented() = 1;
help?> undocumented
search: undocumented

  No documentation found.

  undocumented is of type Function:

  Summary:

  type Function <: Any

  Fields:

  fptr :: Ptr{Void}
  env  :: Any
  code :: LambdaStaticData
```

Maybe just print the method table?

**Comment by jakebolewski** [comment]
```julia
julia> f = (x::Int) -> nothing
(anonymous function)

help?> f
search: f fd for fma fld fft full find filt fill fft! fdio frexp foldr foldl flush floor float first findn filt! fill! fetch FFTW flipud fliplr fldmod findnz

  No documentation found.

  f is an anonymous Function.

  (x::Int) -> ...

julia> g() = nothing
g (generic function with 1 method)

julia> g(x) = nothing
g (generic function with 2 methods)

julia> g(x,y) = nothing
g (generic function with 3 methods)

help?> g
search: g gc get gcd get! gcdx gperm gamma global golden givens getpid getkey gensym GenSym getfield gradient getindex GotoNode graphemes getipaddr gc_enable

  No documentation found.

  g is a generic Function.

  # 3 methods for generic function "g":
  g() at none:1
  g(x) at none:1
  g(x, y) at none:1
```

this doesn't handle cases like `help?> g(1)` but I'm done.

**Comment by MichaelHatherly** [comment]
``Any`` and ``ANY`` aren't the same thing.

```
julia> Any === ANY
false
```

 ``ANY`` can be used to stop Julia from specialising on a particular argument type, or at least that's what I believe that it does... See the difference in output of ``@code_llvm`` for the following functions:

```jl
f(x::Any) = -x
g(x::ANY) = -x

@code_llvm f(1)
@code_llvm g(1)
```

> Where is type ``ANY`` defined

Looks like it's here:

https://github.com/JuliaLang/julia/blob/17e36e6bb3cdb3a9dfd1ef84c259f19cc1d934d1/src/jltypes.c#L3490
https://github.com/JuliaLang/julia/blob/17e36e6bb3cdb3a9dfd1ef84c259f19cc1d934d1/src/builtins.c#L1297

---

Really liking the new output @jakebolewski, thanks!

**Comment by tkelman** [comment]
backported in #13072

**Pull request merged**